crates/hir-ty/src/infer.rs

```diff
@@ -825,13 +830,13 @@ impl<'a> InferenceContext<'a> {
                 self.write_binding_ty(self_param, ty);
             }
         }
-        let mut params_and_ret_tys = Vec::new();
+        let mut tait_candidates = FxHashSet::default();
         for (ty, pat) in param_tys.zip(&*self.body.params) {
             let ty = self.insert_type_vars(ty);
             let ty = self.normalize_associated_types_in(ty);
 
             self.infer_top_pat(*pat, &ty);
-            params_and_ret_tys.push(ty);
+            tait_candidates.insert(ty);
```

Comment:
```suggestion
			if ty.data(Interner).flags.intersects(TypeFlags::HAS_TY_OPAQUE) {
	            tait_candidates.insert(ty);
	        }
```
(just noticed the flag exists)

Reply:
It's strange that this makes the tests fail 🤔 

```
failures:

---- tests::diagnostics::no_mismatches_on_atpit stdout ----
thread 'tests::diagnostics::no_mismatches_on_atpit' panicked at crates/hir-ty/src/tests/diagnostics.rs:102:5:
Unexpected type mismatches:
257..259: expected impl Sized, got ()

note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- tests::type_alias_impl_traits::associated_type_impl_trait stdout ----
thread 'tests::type_alias_impl_traits::associated_type_impl_trait' panicked at crates/hir-ty/src/tests/type_alias_impl_traits.rs:7:5:
Unexpected type mismatches:
203..205: expected impl Foo + ?Sized, got S1


---- tests::type_alias_impl_traits::associated_type_impl_traits_complex stdout ----
thread 'tests::type_alias_impl_traits::associated_type_impl_traits_complex' panicked at crates/hir-ty/src/tests/type_alias_impl_traits.rs:35:5:
Unexpected type mismatches:
293..295: expected impl Foo + ?Sized, got S1
585..587: expected impl Foo + ?Sized, got S1
595..597: expected impl Bar + ?Sized, got S2



failures:
    tests::diagnostics::no_mismatches_on_atpit
    tests::type_alias_impl_traits::associated_type_impl_trait
    tests::type_alias_impl_traits::associated_type_impl_traits_complex

test result: FAILED. 783 passed; 3 failed; 0 ignored; 0 measured; 0 filtered out; finished in 3.67s
```

Reply:
Oh, I see. It seems that the types we are inserting has some inference vars inside and

https://github.com/rust-lang/rust-analyzer/blob/c9955bf86bd7fb779c44048ef67e8bf0fec76291/crates/hir-ty/src/infer.rs#L949

this line is adding opaque types when the top type doesn't have the flag `HAS_TY_OPAQUE`

Reply:
I guess we can check for either `HAS_TY_OPAQUE` or `HAS_TY_INFER` then? If not feel free to merge as is

Reply:
Reasonable alternative 👍